### PR TITLE
enable jit+pmap by merging pxla.py and xla.py

### LINKED ISSUE #804: xla in pmap fails (i.e. jit-of-pmap or lax.scan with collectives)
The parallel xla interpreter currently doesn't properly support nested jit compilation. 
A practical example of this issue is when trying to use psum from within scan:

```
pmap( partial(lax.scan, lambda c, x: (c + lax.psum(x, "i") , c), 0.), axis_name="i" )(np.ones((8, 4)))
```

Scan need to compiles the body of the loop using xla which fails because psum is only defined in the context of the pxla interpreter.

### PR DESCRIPTION
**Pull request by mattjj**
This change is essentially de-duplicating the XLA lowering logic between xla.py and pxla.py. Only the latter was capable of handling collectives (aka pmap primitives), which meant that these didn't work:

1. some compositions of jit and pmap, like jit-of-pmap
2. collectives inside initial-style control flow like scan
3. jax.xla_computation on a function involving collectives

By merging the logic into xla.py, now all the lowering machinery works with everything. Woo!

The pxla.py file still exists and contains mostly dynamic/runtime components for pmap and functions used only by pmap and collectives translations. In particular, pxla.py has

* the pmap impl, particularly the dispatching logic for top-level pmaps, including argument sharding and lazy sharded result persistence
* the ShardedDeviceArray / ShardedDeviceTuple classes
* the dynamic (trace-time) axis environment data structures and logic and the special axis_index primitive
* the split-axis transformation for soft_pmap
* PmapPrimitive (just a tagged version of Primitive)
* the static sharding/unsharding logic for pmap-inside-jit/pmap

These things moved over to xla.py:

* the logic for lowering pmap primitives, especially the static axis environment used during xla lowering

This change refactors the translation rule tables a bit. Instead of just having two tables, there are now five, and they contain rules with slightly different type signatures:
* the `translations` table has rules with the same signatures as always, i.e. `CompBuilder -> [XlaOperand] -> ParamsDict -> XlaOperandOut`
* the `backend_specific_translations` table is keyed by platform name strings and has dict values that each have the same type as `translations`
* the `parallel_translations` table is used for primitives modeling parallel collectives, and so it has rules with signature `CompBuilder -> [XlaOperand] -> ReplicaGroups -> ParamsDict -> XlaOpOut`
* the `initial_style_translations` table is for the initial-style control flow primitives (like `scan`), for which the translation rules themselves lower jaxprs to XLA computations and thus require the static axis env to be passed in; the rules there have signature `CompBuilder -> AxisEnv -> [XlaOperand] -> ParamsDict -> XlaOpOut`
* the `call_translations` table is used for `xla_call` and `xla_pmap`, i.e. the primitives underlying `jit` and `pmap` respectively, and has rules with signature `CompBuilder -> Jaxpr -> AxisEnv -> [XlaOp] -> [XlaOp] -> ParamsDict -> XlaOp`

Having these as separate tables is an uninteresting implementation detail. The lowering function `_jaxpr_computation` just does a case analysis on whether the primitive being translated has an entry in any table (where the `backend_specific_translations` table must be checked before the `translations` table, since some primitives may be entered in both).

This change fixes #804 also addresses #852, in that the lax control flow impls for those primitives are now based on Python-level jaxpr interpreters rather than XLA compilation, but we should probably wait to close the latter issue until we benchmark and improve things more. This change at least seems not to be a performance regression: on my machine the lax control flow tests go from running in ~20s to running in ~14s.

This change also adds a docstring for `jax.xla_computation` and some basic tests.

### CODE REVIEW

**Comment by mattjj** [comment]
Internal tests pass!
